### ISSUE TITLE
Bug: `for (using foo in {});` should be invalid

### ISSUE THREAD

**Issue by fisker** [typescript-eslint/typescript-eslint#7555]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.0-beta&showAST=es&fileType=.tsx&code=GYewTgBAFArgzgSwHYHMKhBZEDeBfASgG4g&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
for (using foo in {});
```

- [TS Playground](https://www.typescriptlang.org/play?#code/GYewTgBAFArgzgSwHYHMKhBZEDeBfASgG4g)
- [Prettier Playground (babel)](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAzCAnABACgK4DOAllAOZboRYlbAC+AlANwgA0IEADjEdAcqACGGDBADuABWEJ+KQQBsxggJ792AIwyCwAazgwAyoIC2cADIk4yVAoJwNW3foOdtJUshgY89kHeNEnt6+xGTycACKeBDw1ra+AFYEAB4G7uFRMVZINvJ27ACO0fASopyyIIIEALRQcHAAJo1sIF6CRPLuAMIQxsaCyJXy8i2hpOEAgjBeROp4JXAYFnVxeb4AFjDG8gDq60TwBK5gcAYyB0QAbgfKg2AEaiCXPgCSUE2wBmAYRNwT7wYYMpwqt8iBOKI7DstJxBhC4HYMJcrOwSIiYKVBKR+qDfK4MIjBupBOo4CN2BCSDAdkQGjB1sgABwABnYGDgRSI7Mx2IGOXi7BgJJpdIZSAATOxCHAACok2S5MFwYykhpNBpmQRkPBYuAAMUw-Wm7kGgnmEBAdDoQA)

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Error thrown.

### Actual Result

Successfuly parsed.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `5.6.0` |

**Comment by fisker** [comment]
The check can be added here. https://github.com/typescript-eslint/typescript-eslint/blob/d2973cca391eeb763d72c5ef2f49cdafe4ffc216/packages/typescript-estree/src/convert.ts#L920

**Comment by Solo-steven** [comment]
HI @fisker  thanks for your reply, but is the Typescript Program instance already doing the semantic diagnosis for us, why not just use but manual check in the coverter?

**Comment by bradzacher** [comment]
It costs around 50ms per file to calculate diagnostics. This adds up quickly across any non-trivial codebase and so we purposely avoid it.

Instead we implement our own restrictions within our AST conversion as we can do it very quickly during the traversal - we have to inspect and disect the AST any way to convert it so it's easy and natural for us to enforce certain invariants during this process.
